docs/testing-principles.md

```diff
@@ -0,0 +1,200 @@
+orgs for tests:
+
+```
+@pytest.mark.vcr()
+@pytest.mark.org_shape("person_accounts", "config_person_accounts")
+def test_some_person_account_feature(self):
+    ...
+```
+
+This will run the `config_person_accounts` flow against the `person_accounts`
+org shape and run the test in the context of that org. At least that's
+the goal. This infrastructure has not been tested or used much yet (e.g. for
+actual person accounts testing).
+
+### Simulating Networking Errors with VCR
+
+It is possible to simulate networking errors by hand-editing VCR files.
+This pattern can be observed [here](https://github.com/SFDO-Tooling/CumulusCI/blob/35b07ebfa80170f6e0e9d8cabd3da75bf41d5d59/cumulusci/utils/salesforce/tests/test_count_sobjects.py#L23) and [here](https://github.com/SFDO-Tooling/CumulusCI/blob/e3ba7e8b542755b6395dbc895daee3d1ea3d11dc/cumulusci/utils/salesforce/tests/cassettes/ManualEdit_TestCountSObjects.test_count_sobjects__network_errors.yaml#L16).
+
+### Service Connectors
+
+Third party services (including Salesforce) should always be wrapped in Service Connector
+abstractions.
+
+These abstractions can be tested with Requests and VCR.
+
+The overall business logic can be tested by injecting fakes for Service Connectors.
+
+A client library like simple_salesforce or github3 is not necessarily a good service
+connector abstraction, because they work at a protocol level and not a business logic
+level. Two examples, from simple_salesforce:
+
+1. We make heavy use of simple_salesforce.restful(), which tunnels MANY different types
+   of requests through a single endpoint and is therefore hard to fake.
+
+2. We make heavy use of simple_salesforce.query() which has the same problem.
+
+Github3 is better, but arguably if we had started using it by wrapping it we'd be in a
+better position to add other VCS regardless, so maybe we should just always wrap
+client libraries by default. The case of Github3 is argably, but simple_salesforce
+is not: the fact that we do all sorts of protocol-level things (passing URLs,
+fiddling headers) is a clear giveawaay that we should have wrapped it.
+
+#### Wrap Universal Interfaces
+
+An abstraction like Salesforce.query() should be used for situations such as the
+end-user passing queries to us to execute. We, internally, should always wrap
+it in a business-logic layer.
+
+For example, instead of:
+
+```python
+sf_org.query("SELECT LicenseDefinitionKey FROM UserLicense")
+```
+
+There should be:
+
+```
+sf_org.get_user_licenses()
+```
+
+The latter can be faked much more easily and scalably. Imagine if you are testing
+a large component which calls `sf_org.query()` six (or six hundred!) times. You'll
+need to use string matching (which is intrinsically fragile) to figure out which
+return values to return for which call.
+
+Between the benefits of simpler tests, cleaner code and easier replacement of abstractions,
+ServiceConnectors will pay for themselves quickly.
```

Comment:
These two sections seem like a very persuasive ADR in favor of using the Service Connector pattern.